Made hook.Add/Remove not silently fail on arg type mismatch

Kefta opened the pull request:
In the rare case a typo was made for the hook args, it is now noticeable.

neico commented on garrysmod/lua/includes/modules/hook.lua:
```diff
@@ -41,7 +44,7 @@ end
 function Remove( event_name, name )
 
-	if ( !isstring( event_name ) ) then return end
+	if ( !isstring( event_name ) ) then error( "bad argument #1 to 'Add' (string expected, got " .. type( event_name ) .. ")" ) end
```
'Add' -> 'Remove'

Kefta replied:
Thanks!

neico commented:
What did those first lines do to you to deserve this much bashing? (alas, why the space changes?)

Kefta commented:
Each of them was a space; I don't understand why considering they weren't aligned, and everything else in GMod uses tab alignment, thus I just removed them since I added new locals anyway.

Kefta commented:
Should the second argument be type-checked? The wiki says the identifier can be a string, table, or entity/panel, and that numbers/bools aren't allowed, but everything is allowed right now.

robotboy655 commented:
Please add newlines, and perhaps it still should be error() because this doesn't look terribly helpful:
![image](https://user-images.githubusercontent.com/3299036/58981188-9097d280-87da-11e9-8b60-157fc18919a9.png)

Or at least put the `return`s back in so it functions like it used to.

thegrb93 commented:
Why doesn't ErrorNoHalt provide the stacktrace?

Kefta commented:
I'm thinking of PRing an ErrorNoHaltStack function.

robotboy655 commented:
I am afraid adding stack trace to `ErrorNoHalt` will cause problems for existing code, especially since `ErrorNoHalt` doesn't add its own new line.

Kefta commented:
It would be a separate function, literally called ``ErrorNoHaltStack``. Here's my first stab at it:
```
function ErrorNoHaltStack(sMessage, uLevel --[[= 1]])
	if (not isstring(sMessage)) then
		error(string.format("bad argument #1 to 'ErrorNoHaltStack' (string expected, got %s)", type(sMessage)), 2)
	end
	
	if (uLevel == nil) then
		uLevel = 2
	elseif (isnumber(uLevel)) then
		uLevel = uLevel + 1
	else
		error(string.format("bad argument #2 to 'ErrorNoHaltStack' (number expected, got %s)", type(uLevel)), 2)
	end
	
	local tInfo = debug.getinfo(uLevel, "nS")
	
	if (tInfo == nil) then
		ErrorNoHalt(string.format("[ERROR] %s\n", sMessage))
		
		return
	end
	
	local sSource = tInfo.source
	
	if (sSource == "[C]") then
		sMessage = string.format("[ERROR] %s\n", sMessage)
	else
		sMessage = string.format("[ERROR] %s:%i: %s\n", sSource, tInfo.lastlinedefined, sMessage)
	end
	
	local iStackLevel = 1
	
	::StackLoop::
	
	sMessage = string.format("%s%u. %s - %s:%i\n", sMessage, string.rep(" ", iStackLevel + 1), iStackLevel, tInfo.name, tInfo.short_src, tInfo.lastlinedefined)
	uLevel = uLevel + 1
	tInfo = debug.getinfo(uLevel, "nS")
	
	if (tInfo ~= nil) then
		iStackLevel = iStackLevel + 1
		
		goto StackLoop
	end
	
	ErrorNoHalt(sMessage)
end
```

Might be better to add in C but this has the benefit of actually altering the stack trace along with the error level argument unlike ``error``.

thegrb93 commented:
you can `repeat until` instead of goto might be preferred but lgtm

Bo98 commented on garrysmod/lua/includes/modules/hook.lua:
```diff
@@ -22,11 +25,11 @@ function GetTable() return Hooks end
-	if ( !isfunction( func ) ) then return end
-	if ( !isstring( event_name ) ) then return end
+	if ( !isstring( event_name ) ) then ErrorNoHalt( "bad argument #1 to 'Add' (string expected, got " .. type( event_name ) .. ")" ) end
+	if ( !isfunction( func ) ) then ErrorNoHalt( "bad argument #3 to 'Add' (function expected, got " .. type( func ) .. ")" ) end
```
If you are using `ErrorNoHalt` then you probably want the `return` still there to prevent further errors.

Kefta commented:
This PR now depends on https://github.com/Facepunch/garrysmod/pull/1587.
